### PR TITLE
add eliminate duplicate initializer pass

### PR DESCRIPTION
**Pull request by fumihwh**

### CODE REVIEW

File: onnxoptimizer/passes/eliminate_duplicate_initializer.h [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        graph.eraseInitializerAndInput(j_value);                             \
+        initializers_removed++;                                              \
+      }                                                                      \
+    }                                                                        \
+  }
+        switch (i_tensor.elem_type()) {
+          case ONNX_NAMESPACE::TensorProto_DataType_FLOAT: {
+            DO_COMPARISON(float)
+            break;
+          }
```
Comment (daquexian) [inline onnxoptimizer/passes/eliminate_duplicate_initializer.h:125]: how about defining a new macro like
```cpp
#define CASE_DO_COMPARISON(ONNX_DTYPE_SUFFIX, CPP_DTYPE) \
          case ONNX_NAMESPACE::TensorProto_DataType_##ONNX_DTYPE_SUFFIX: { \
            DO_COMPARISON(CPP_DTYPE) \
            break; \
          } 
```

File: onnxoptimizer/passes/eliminate_duplicate_initializer.h [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+namespace ONNX_NAMESPACE {
+namespace optimization {
+
+struct EliminateDuplicateInitializer final : public FullGraphBasedPass {
+  explicit EliminateDuplicateInitializer()
+      : FullGraphBasedPass(
+            PassType::Nop,
+            PassEfficiency::Complete,
+            PassOptimizationType::Compute) {}
```
Comment (daquexian) [inline onnxoptimizer/passes/eliminate_duplicate_initializer.h:27]: Should be `PassOptimizationType::Memory`?

File: onnxoptimizer/passes/eliminate_duplicate_initializer.h (hunk 1 above)
Comment (fumihwh) [reply onnxoptimizer/passes/eliminate_duplicate_initializer.h]: Sure.

File: onnxoptimizer/passes/eliminate_duplicate_initializer.h (hunk 2 above)
Comment (fumihwh) [reply onnxoptimizer/passes/eliminate_duplicate_initializer.h]: Fixed.

File: onnxoptimizer/passes/eliminate_duplicate_initializer.h [hunk 3]
```diff
@@ -89,61 +87,52 @@ struct EliminateDuplicateInitializer final : public FullGraphBasedPass {
...
+      if (std::equal(i_data.begin(), i_data.end(), j_data.begin())) {          \
+        visited.insert(*iter_j);                                               \
+        Value *j_value = input_map.find(j_tensor.name())->second;              \
+        j_value->replaceAllUsesWith(i_value);                                  \
+        graph.eraseInitializerAndInput(j_value);                               \
+        initializers_removed++;                                                \
+      }                                                                        \
+    }                                                                          \
+  }
+#define CASE_DO_COMPARISON(ONNX_DTYPE_SUFFIX, CPP_DTYPE)                       \
```
Comment (daquexian) [inline onnxoptimizer/passes/eliminate_duplicate_initializer.h:117]: `#undef CASE_DO_COMPARISON` looks missing

File: onnxoptimizer/passes/eliminate_duplicate_initializer.h [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    if (iter_j_initializer == graph.initializers().end()) {                    \
+      visited.insert(*iter_j);                                                 \
+      continue;                                                                \
+    }                                                                          \
+    Tensor j_tensor = *iter_j_initializer;                                     \
+    if (i_tensor.elem_type() != j_tensor.elem_type()) {                        \
+      continue;                                                                \
+    } else {                                                                   \
+      const std::vector<data_type> j_data = ParseData<data_type>(&j_tensor);   \
+      if (std::equal(i_data.begin(), i_data.end(), j_data.begin())) {          \
```
Comment (daquexian) [inline onnxoptimizer/passes/eliminate_duplicate_initializer.h:108]: will `memcmp` + an extra data type equality check be better?

**Review by daquexian — APPROVED** [review]
Great!

**Pull request merged**